### DISCUSSION TITLE
(Display) Number Generation

### DISCUSSION THREAD

**Discussion by rjhllr** [medusajs/medusa#965] [Feature Requests]
This idea stems from an [ongoing discussion](https://github.com/medusajs/medusa/discussions/964) with @olivermrbl and @adrien2p talking about the same concept for authentication systems (middlewares) 
Most advanced ecom systems have some kind of a "number ranges" feature. This is often used in big companies where customers registering from the webshop get different customer numbers than e.g. internal customers. This is also often an requirement with migrations of existing shops: when importing the old orders, you often need some way to differ new incoming shop orders (prefix/suffix, different scheme…). Third example: numbers are generated by external systems (e.g. customer numbers for first-time webshop customers come from a CRM system, order numbers coming from ERP). Fourth example would be small/new stores (for obvious reasons) using random or massive-scale stores (for avoiding collisions) using UUIDs. 

I've taken a look at orders out of interest. Currently, the display_id is implicitly generated by not being set in the orderService and then letting the DBMS handle the generation (as I understood it, see `resolveDbGenerationStrategy`, `medusa/src/models/order.ts`). Inserting a strategy that keeps the current behavior (= don't set a value, no-op strategy) would allow developers to very easily modify entity number generation. 

Core change to achieve this for example for orders: 1 line in the orderService, one interface (/abstract class as per @adrien2p) added. Pretty neat.

**Comment by aglitchman** [comment]
Is there an option to change the ID generator now for orders? They go in a **continuous** order from 1... and this makes it easy for our competitors to estimate order volumes.

**Comment by bqst** [comment]
I've successfully implemented a migration to replace Medusa's default sequential display_id with random 7-digit numbers for better order privacy and UX. 
The main issue was that Medusa's order table has a default constraint nextval('order_display_id_seq'::regclass) on the display_id field, which prevented the trigger from working properly since the value was already set before insertion.

```typescript
import { Migration } from "@mikro-orm/migrations"

export class Migration20250701140000 extends Migration {
  override async up(): Promise<void> {
    // 1. First, remove the default value
    this.addSql(`
      ALTER TABLE "order" ALTER COLUMN display_id DROP DEFAULT;
    `)

    // 2. Create the random ID generation function
    this.addSql(`
      CREATE OR REPLACE FUNCTION generate_random_display_id()
      RETURNS INTEGER AS $$
      DECLARE
        random_id INTEGER;
        max_attempts INTEGER := 100;
        attempts INTEGER := 0;
      BEGIN
        LOOP
          -- Generate random 7-digit number between 1000000 and 9999999
          random_id := floor(random() * 9000000 + 1000000)::INTEGER;
          
          -- Check if this ID already exists
          IF NOT EXISTS (SELECT 1 FROM "order" WHERE display_id = random_id) THEN
            RETURN random_id;
          END IF;
          
          attempts := attempts + 1;
          IF attempts >= max_attempts THEN
            RAISE EXCEPTION 'Could not generate unique display_id after % attempts', max_attempts;
          END IF;
        END LOOP;
      END;
      $$ LANGUAGE plpgsql;
    `)

    // 3. Update the existing orders with random display_ids
    this.addSql(`
      DO $$
      DECLARE
        order_record RECORD;
        new_display_id INTEGER;
      BEGIN
        FOR order_record IN SELECT id FROM "order" LOOP
          new_display_id := generate_random_display_id();
          UPDATE "order" SET display_id = new_display_id WHERE id = order_record.id;
        END LOOP;
      END $$;
    `)

    // 4. Create the unique index after updating the existing data
    this.addSql(`
      CREATE UNIQUE INDEX IF NOT EXISTS "IDX_order_display_id_unique" ON "order" (display_id);
    `)

    // 5. Create the trigger function (modified to be more robust)
    this.addSql(`
      CREATE OR REPLACE FUNCTION set_order_display_id()
      RETURNS TRIGGER AS $$
      BEGIN
        -- Seulement si display_id n'est pas déjà défini
        IF NEW.display_id IS NULL THEN
          NEW.display_id := generate_random_display_id();
        END IF;
        RETURN NEW;
      END;
      $$ LANGUAGE plpgsql;
    `)

    // 6. Create the trigger
    this.addSql(`
      DROP TRIGGER IF EXISTS order_display_id_trigger ON "order";
      CREATE TRIGGER order_display_id_trigger
        BEFORE INSERT ON "order"
        FOR EACH ROW
        EXECUTE FUNCTION set_order_display_id();
    `)
  }

  override async down(): Promise<void> {
    // Drop the trigger
    this.addSql(`DROP TRIGGER IF EXISTS order_display_id_trigger ON "order";`)

    // Drop the trigger function
    this.addSql(`DROP FUNCTION IF EXISTS set_order_display_id();`)

    // Drop the random ID generation function
    this.addSql(`DROP FUNCTION IF EXISTS generate_random_display_id();`)

    // Drop the unique index
    this.addSql(`DROP INDEX IF EXISTS "IDX_order_display_id_unique";`)

    // Restore the original DEFAULT constraint
    this.addSql(`
      ALTER TABLE "order" ALTER COLUMN display_id SET DEFAULT nextval('order_display_id_seq'::regclass);
    `)
  }
}
```
